src/util/interval_union.cpp

```diff
@@ -0,0 +1,236 @@
+  while(it1 != intervals.end() && it2 != other.intervals.end())
+  {
+    intervalt intersection = *it1;
+    intersection.intersect_with(*it2);
+    if(!intersection.empty())
+      result.intervals.emplace_back(std::move(intersection));
+    if(it1->upper < it2->upper)
+      ++it1;
+    else
+      ++it2;
+  }
+  return result;
+}
+
+interval_uniont interval_uniont::make_union(const interval_uniont &other) const
+{
+  if(intervals.empty())
+    return other;
+  if(other.intervals.empty())
+    return *this;
+
+  interval_uniont result{};
+  auto it1 = intervals.begin();
+  auto it2 = other.intervals.begin();
+  if(!it1->lower_set)
+  {
+    if(!it2->lower_set)
+    {
+      result = smaller_or_equal(std::max(it1->upper, it2->upper));
+      ++it1;
+      ++it2;
+    }
+    else
+    {
+      result.intervals.push_back(*it1);
+      ++it1;
+    }
+  }
+  else
+  {
+    if(!it2->lower_set)
+    {
+      result.intervals.push_back(*it2);
+      ++it2;
+    }
+    else if(it1->lower <= it2->lower)
+    {
+      result.intervals.push_back(*it1);
+      ++it1;
+    }
+    else
+    {
+      result.intervals.push_back(*it2);
+      ++it2;
+    }
+  }
+  while(it1 != intervals.end() || it2 != other.intervals.end())
+  {
+    intervalt &back = result.intervals.back();
+    INVARIANT(
+      !back.lower_set ||
+        ((it1 == intervals.end() || it1->lower >= back.lower) &&
+         (it2 == other.intervals.end() || it2->lower >= back.lower)),
+      "intervals are processed in order");
+    if(it1 != intervals.end() && it1->lower <= back.upper + 1)
+    {
+      back.approx_union_with(*it1);
+      ++it1;
+    }
+    else if(it2 != other.intervals.end() && it2->lower <= back.upper + 1)
+    {
+      back.approx_union_with(*it2);
+      ++it2;
+    }
+    else if(
+      it1 != intervals.end() &&
+      (it2 == other.intervals.end() || it1->lower <= it2->lower))
+    {
+      result.intervals.push_back(*it1);
+      ++it1;
+    }
+    else
+    {
+      result.intervals.push_back(*it2);
+      ++it2;
+    }
+  }
+  return result;
+}
+
+bool interval_uniont::is_empty() const
+{
+  return intervals.empty();
+}
+
+optionalt<mp_integer> interval_uniont::maximum() const
+{
+  if(intervals.empty())
+    return {};
+  const intervalt &back = intervals.back();
+  if(back.upper_set)
+    return back.upper;
+  return {};
+}
+
+optionalt<mp_integer> interval_uniont::minimum() const
+{
+  if(intervals.empty())
+    return {};
+  const intervalt &front = intervals.front();
+  if(front.lower_set)
+    return front.lower;
+  return {};
+}
+
+std::string interval_uniont::to_string() const
+{
+  if(intervals.empty())
+    return "[]";
+  std::ostringstream output;
+  for(const intervalt &i : intervals)
+  {
+    output << '[';
+    if(i.lower_set)
+      output << i.lower;
+    output << ':';
+    if(i.upper_set)
+      output << i.upper;
+    output << "]";
+  }
+  return output.str();
+}
+
+exprt interval_uniont::contains_expr(const exprt &e) const
```

Comment:
`get_contains_expr`? `contains_expr` suggests it would return a boolean indicating whether `to_constant_expr(e)` fell within the specified interval

Reply:
`get_contains_expr` makes me think of a getter and not something that is constructing an expression. `contains_expr` follows the model of `equal_expr` which doesn't return a Boolean but an expression, but I agree this can be ambiguous, however the return type makes it clear.

Reply:
`make_contains_expr`?